File: LeetCode/Study/ArrayAndString/18.YangHuiTriangle_II.py
"""
给定一个非负索引 rowIndex，返回「杨辉三角」的第 rowIndex 行。
在「杨辉三角」中，每个数是它左上方和右上方的数的和。
"""

from typing import List

class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        row = [1]
        for i in range(1, rowIndex + 1):
            # 逆序更新，避免覆盖
            for j in range(i - 1, 0, -1):
                row[j] = row[j] + row[j - 1]
            row.append(1)
        return row

def test():
    sol = Solution()
    test_cases = [
        (0, [1]),  # 第 0 行
        (1, [1, 1]),  # 第 1 行
        (2, [1, 2, 1]),  # 第 2 行
        (3, [1, 3, 3, 1]),  # 第 3 行
        (4, [1, 4, 6, 4, 1]),  # 第 4 行
        (5, [1, 5, 10, 10, 5, 1]),  # 第 5 行
    ]
    
    for i, (rowIndex, expected) in enumerate(test_cases):
        result = sol.getRow(rowIndex)
        print(f"Test case {i + 1}:")
        print(f"Input: {rowIndex}")
        print(f"Expected: {expected}, Result: {result}")
        print("✅ Pass" if result == expected else "❌ Fail")
        print("-" * 40)

if __name__ == "__main__":
    test()

File: LeetCode/Study/ArrayAndString/18.YangHuiTriangle_II.java
/*
 * 给定一个非负索引 rowIndex，返回「杨辉三角」的第 rowIndex 行。
 * 在「杨辉三角」中，每个数是它左上方和右上方的数的和。
 */


import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> getRow(int rowIndex) {
        List<Integer> row = new ArrayList<>();
        row.add(1); // 第一行的第一个元素是 1
        for (int i = 1; i <= rowIndex; i++) {
            // 从后向前更新行，避免覆盖前一个元素
            for (int j = i - 1; j > 0; j--) {
                row.set(j, row.get(j) + row.get(j - 1));
            }
            // 添加新的元素到行的末尾
            row.add(1);
        }
        return row;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        int[] testInputs = {0, 1, 2, 3, 4, 5};
        List<List<Integer>> expectedOutputs = List.of(
            List.of(1),
            List.of(1, 1),
            List.of(1, 2, 1),
            List.of(1, 3, 3, 1),
            List.of(1, 4, 6, 4, 1),
            List.of(1, 5, 10, 10, 5, 1)
        );

        for (int i = 0; i < testInputs.length; i++) {
            int rowIndex = testInputs[i];
            List<Integer> expected = expectedOutputs.get(i);
            List<Integer> result = sol.getRow(rowIndex);

            System.out.println("Test case " + (i + 1) + ":");
            System.out.println("Input: " + rowIndex);
            System.out.println("Expected: " + expected);
            System.out.println("Result:   " + result);
            System.out.println(expected.equals(result) ? "✅ Pass" : "❌ Fail");
            System.out.println("----------------------------------------");
        }
    }
}